taskgroup.go

```diff
@@ -42,70 +39,55 @@ func New(ef ErrorFunc) *Group { return &Group{onError: ef} }
 func (g *Group) Go(task Task) *Group {
 	g.wg.Add(1)
 	g.init()
-	errc := g.errc
 	go func() {
 		defer g.wg.Done()
 		if err := task(); err != nil {
-			errc <- err
+			g.handleError(err)
 		}
 	}()
 	return g
 }
 
-func (g *Group) init() {
-	// The first time a task is added to an otherwise clear group, set up the
-	// error collector goroutine.  We don't do this in the constructor so that
-	// an unused group can be abandoned without orphaning a goroutine.
-	g.setup.Do(func() {
-		if g.onError == nil {
-			g.onError = func(e error) error { return e }
-		}
-		g.err = nil
-		g.edone = make(chan struct{})
-		g.reset = sync.Once{}
-
-		errc := make(chan error)
-		g.errc = errc
-		go func() {
-			defer close(g.edone)
-			for err := range errc {
-				e := g.onError(err)
-				if e != nil && g.err == nil {
-					g.err = e // capture the first error always
-				}
-			}
-		}()
-	})
+func (g *Group) handleError(err error) {
+	g.μ.Lock()
+	defer g.μ.Unlock()
+	e := g.onError.filter(err)
+	if e != nil && g.err == nil {
+		g.err = e // capture the first unfiltered error always
```

Comment:
unrelated thought: now that errors.Join exists, should this accumulate all errors and return them as a wrapped join? Not for this PR of course.

Reply:
I thought about that, but I didn't want to keep around an unbounded slice of error values. The caller can choose to accumulate them, but not to "un-accumulate" them.